### ISSUE TITLE
Unicode normalization: Enable by default for speech and reconsider character navigation

### ISSUE THREAD

**Issue by LeonarddeR** [nvaccess/nvda#16616]
### Is your feature request related to a problem? Please describe.
Recently, Unicode normalization has been added to NVDA, but in #16466, there was some discussion about whether it should be enabled by default.
Furthermore, there was some debate about character navigation, i.e. whether normalization should be enabled by default for character navigation, and/or whether the delayed character description feature could be utilized to some extend.

### Describe the solution you'd like
1. Enable normalization by default for speech output.
2. When enabled, always normalize (i.e. also with character navigation) but prepend the word 'normalized' or something less technical to the character.

### Describe alternatives you've considered
@CyrilleB79 proposed a three choice option in https://github.com/nvaccess/nvda/issues/16466#issuecomment-2132149218:

1. Normalization disabled (Current default)
2. Normalization enabled to read text but disabled to read by character (current behavior of enabled)
3. Normalization enabled for all (not yet implemented)

I'm happy to consider this alternative as well, though it will probably take more effort. I'm now at the point where I think if we add a proper term prepended for character navigation, that would be the best solution.

### Additional context
1. I'm reluctant to take the character descriptions feature into account:

    * Current use of these is for phonetics. I'm inclined to say that adding Unicode character descriptions is a fundamental change to this feature
    * There are no translated names for the several mathematical Unicode chars in CLDR

2. Braille is beyond discussion.

### Discussion guidelines
When attending to the discussion, please avoid repeating standpoints over again that were already stated in https://github.com/nvaccess/nvda/issues/16466. It is clear that we have different opinions and that is okay.

**Comment by LeonarddeR** [comment]
Feel free to chime in, @Adriani90, @XLTechie and others.

**Comment by Adriani90** [comment]
I agree in principle with this proposed solution, thanks for considering this @LeonarddeR.
One thing though:
> 2. When enabled, always normalize (i.e. also with character navigation) but prepend the word 'normalized' or something less technical to the character.

When reading this example pdf, exploring the equations by character will very probably be needed quite often, this will be the case in similar documents as well. It feels quite inefficient to hear "normalized" or something like that before every character. Could we agree in making this part optional as well? Like a conditional checkbox? If normalization is on, user can choose whether "normalized" should be pronounced before every character or not.

Example pdf:
https://www.andrews.edu/~rwright/physics/Physics%20Formula%20Sheet.pdf

**Comment by Adriani90** [comment]
Another alternative for character by character navigation could be to speak the normalized characters in a lower voice pitch.

**Comment by LeonarddeR** [comment]
> When reading this example pdf, exploring the equations by character will very probably be needed quite often, this will be the case in similar documents as well. It feels quite inefficient to hear "normalized" or something like that before every character. Could we agree in making this part optional as well? Like a conditional checkbox? If normalization is on, user can choose whether "normalized" should be pronounced before every character or not.

Makes sense, we have the same for caps.

> Another alternative for character by character navigation could be to speak the normalized characters in a lower voice pitch.

I'm not sure how to handle normalized capitals in that case.

**Comment by CyrilleB79** [comment]
There are two cases when reading by character:
1. You expect to have math characters at a specific place, e.g. in Word equations. In this case, hearing "normalized" before each character makes character by character reading unusable in Word equations, as also explained by @Adriani90 .
2. You do not expect to have math characters at a specific place, e.g. when reading a nickname. In this case, it may be interesting to hear that the used characters are not normal ones.

Ideally, both cases need to be configurable. And this should be separately configurable from normal reading, or make the 3-choice option that I had suggested in #16466. I acknowledge though that it's more work. If you prefer make something simpler in a first time, I'm fine with this. If the need for a more complex option is confirmed after alpha/beta/stable testing, we can improve the option / the feature later. In the meantime, it's still possible to disable normalization and to enable it by profile when needed, e.g. in Word equations.

Also, I have not written it explicitly in #16466. But, even if it's not my personal choice, I'm fine with enabling normalization by default if you think that it helps the majority.

One last question: If we enable normalization for speech by default, may it be confusing if braillez normalization is not enabled by default at the same time? I have no personal opinion on this; just asking for completeness.

**Comment by CyrilleB79** [comment]
Re pitch modification: that's not a good idea IMO. It may be too difficult to find differences between this 4 possibilities:
* no cap, not normalized
* cap, not normalized
* no cap, normalized
* cap, normalized

By the way, if you stick with adding "normalized" word before the character, put it after the character rather than before. In English, "X normalized" is not grammatically correct, but the most important information is the name of the character, then the fact that it is normalized. In comparison with cap:
* "cap" is a quite short word so having it before the letter may not be too annoying
* in many languages such as French, the translation of "cap" is after the letter, which is much better

**Comment by LeonarddeR** [comment]
> There are two cases when reading by character:
> 
>     1. You expect to have math characters at a specific place, e.g. in Word equations. In this case, hearing "normalized" before each character makes character by character reading unusable in Word equations, as also explained by @Adriani90 .
> 
>     2. You do not expect to have math characters at a specific place, e.g. when reading a nickname. In this case, it may be interesting to hear that the used characters are not normal ones.

I'd personally consider config profiles enough for this.

> Ideally, both cases need to be configurable. And this should be separately configurable from normal reading, or make the 3-choice option that I had suggested in #16466. I acknowledge though that it's more work. If you prefer make something simpler in a first time, I'm fine with this.

I think we need an unassigned gesture to toggle normalization anyway.
Would you agree with a default where:

1. Normalization is enabled but it can be disabled
2. NVDA says "normalized" when doing character by character navigation but that can be disabled separately (like with speaking cap for cappitals)
3. Normalization applies to all speech (i.e. if you want unnormalized character nav) you will need to disable normalization entirely. I think that this desire will be pretty rare, since normalized character nav will be able to announce normalization and the user can use character reporting to report the unnormalized character.

**Comment by CyrilleB79** [comment]
> I think we need an unassigned gesture to toggle normalization anyway.

I fully agree (also for braille by the way). My opinion is that if you need to modify only one setting option on a regular basis, an unassigned keyboard shortcut is more handy than creating a specific profile for that. So almost any new option that is created in settings dialogs (except Advanced panel) should have its corresponding unassigned (or assigned) gesture.

> Would you agree with a default where:
> 
> 1. Normalization is enabled but it can be disabled
> 2. NVDA says "normalized" when doing character by character navigation but that can be disabled separately (like with speaking cap for cappitals)
> 3. Normalization applies to all speech (i.e. if you want unnormalized character nav) you will need to disable normalization entirely. I think that this desire will be pretty rare, since normalized character nav will be able to announce normalization and the user can use character reporting to report the unnormalized character.

I am a bit concerned with point 2; I fear that this will clutter the GUI with a very rarely used option. Why not only keep points 1. and 3 in a first time? I.e. two options:
* no normalization at all
* all normalized (text and character), with no addition of "normalized" information when navigating by character.

People like me who want to know that a character is normalized may just keep normalization off generally and enable it with a gesture when needed. We may evaluate the feature this way and add another option if we get negative feedback.

To summarize I suggest to:
* enable normalization by default
* report normalized character, and just them, when spelling or navigating character by character without reporting "normalized" in any case
* implement additional options or features after a test period according to the feedback.

**Comment by LeonarddeR** [comment]
> To summarize I suggest to:
> 
>     * enable normalization by default
> 
>     * report normalized character, and just them, when spelling or navigating character by character without reporting "normalized" in any case
> 
>     * implement additional options or features after a test period according to the feedback.

How does this relate to your previous comment in https://github.com/nvaccess/nvda/issues/16466#issuecomment-2125638455? Note that you're suggesting now that a speech user by default has no way to find out whether he's dealing with a normal or normalized character, unless reporting the example/numeric form. I thought that like me, you were against this.

**Comment by Adriani90** [comment]
We just need two checkboxes for speech here in my view, and it does not really clutter the GUI:
1. Normalize unicode character checkbox (enabled by default)
2. Press tab: Report "normalized" on every normalized character when navigating character by character checkbox (enabled by default), this checkbox is not available in the gui when 1. is disabled.

This would give enough confort for everyone, and we don't have to deal with a 3 settings combo box.

Still, if you really need to hear the unicode entity of the character
1. You can disable normalization completely or
2. You can press numpad 2 regardless whether the normalization feature is on or off.

**Comment by Adriani90** [comment]
Note that we have already conditional approaches in NVDA, e.g. different voice settings for different synth drivers, or the disclaimer check box in advanced settings. So it is not something uncommon.

**Comment by XLTechie** [comment]
I don't think @CyrilleB79's three-mode proposal is really viable, given the kind of use cases that have been pointed out, and with the end goal of not confusing users over much.
  I believe normalization options for general reading, and for characters, need to be two separate choice lists with their own defaults and their own range of choices.

So, I agree with @LeonarddeR's nuanced proposal above.

IMO the normalization on/off checkbox, as @Adriani90 described, is the best initial choice.

For characters: another checkbox for normalize/don't normalize, might be okay. But I suspect we will ultimately need something like a combo for "don't normalize"/"normalize with announcement"/"normalize without announcement".

(Or whatever word replaces "normalize", of course.)

**Comment by seanbudd** [comment]
@LeonarddeR - can you open a PR to enable this by default when we begin 2025.1 development?

**Comment by LeonarddeR** [comment]
Sure, will do

**Comment by ruifontes** [comment]
I do not agree with this option is enabled by default.
By instance, "nº" should be read as the portuguese word for number, número, and it is read as "no"...

**Comment by LeonarddeR** [comment]
I think @ruifontes raises a valid concern here.

@Adriani90 wrote:
> If you want another pronounciation, you can specify it in the symbols.dic.

Actually with the `º` as in `nº`, it is defined in symbols.dic, but the symbol is always preserved. Therefore the symbol is always processed for normalization, except when navigating by character.
I currently can't think of a way to work around this.

**Comment by XLTechie** [comment]
Can a voice dictionary intercept the resulting normalized word?

If so, perhaps the dictionary processing mechanism could be caused to recognize when a word has been normalized, and apply special rules?

This is a quarter-baked idea.

**Comment by LeonarddeR** [comment]
Normalization is applied to entire strings in speech sequences. This could potentially be very long strings.

**Comment by CyrilleB79** [comment]
Could we add at least in the "Known issues" of this PR what is described in https://github.com/nvaccess/nvda/issues/16616#issuecomment-2297700026, i.e. that the normalization may have unwanted effect in some situation / with some characters (e.g. "nº", common indices or exponents in math expressions such as "ⁱ" or "ₙ")

Moreover, I wonder if we could add a third option for normalization which would be "When suitable", forcing normalization in Word equations and other situations where it would be required and keeping it disabled in other cases. But that would be for another PR in any case.

**Comment by Adriani90** [comment]
I cannot reproduce the math related issue. Where can you reproduce that?Von meinem iPhone gesendetAm 04.09.2024 um 08:29 schrieb Cyrille Bougot ***@***.***>:﻿
Could we add at least in the "Known issues" of this PR what is described in #16616 (comment), i.e. that the normalization may have unwanted effect in some situation / with some characters (e.g. "nº", common indices or exponents in math expressions such as "ⁱ" or "ₙ")
Moreover, I wonder if we could add a third option for normalization which would be "When suitable", forcing normalization in Word equations and other situations where it would be required and keeping it disabled in other cases. But that would be for another PR in any case.

**Comment by SaschaCowley** [comment]
@Adriani90 
> I cannot reproduce the math related issue. Where can you reproduce that?

In Microsoft Word,:
1. Hit `alt+=` to start an equation.
2. Enter `x^3 \times y`.
3. Read the formatted expression with and without unicode normalization enabled.
    - Without: "letter 1-D-4-6-5 three times letter 1-d-4-6-6"
    - With: "x three times y"

Note that I'm using eSpeak as my speech synthesizer.

**Comment by Adriani90** [comment]
@SaschaCowley I see. However, the exponent sign depends on the symbol level of this ^ symbol and not whether normalization is on or off. But maybe I misunderstood something here.


Anyway, I agree a when suitable option would be good to have, however I don't see really a use case where disabling normalization makes things much better.

**Comment by XLTechie** [comment]
Agreed with @Adriani90. I'm not clear on what the math example is trying to prove.

**Comment by SaschaCowley** [comment]
@Adriani90 
> Anyway, I agree a when suitable option would be good to have, however I don't see really a use case where disabling normalization makes things much better.

I absolutely agree with you there.

@XLTechie 
> I'm not clear on what the math example is trying to prove.

That was a misunderstanding on my part. Note that the example math symbols shared by @CyrilleB79 are a superscript 'i' (ⁱ) and a subscript 'n' (ₙ). With unicode normalization disabled, when reading by character, this information is presented. However, with normalization on I just hear "i normalized" or "n normalized".
